### PR TITLE
Inject existing object into MEF2

### LINKED ISSUE #29400: Proposal: Inject existing object into MEF2
MEF1 has method `ComposeExportedValue<T>(T exportedValue)`. MEF2 should have a similar functionality in `System.Composition.Hosting.ConfigurationContainer` (as discussed in dotnet/runtime#18624 and dotnet/runtime#15362)

**Suggested API**

```
public ContainerConfiguration WithInstance<TExport> (TExport instance);
public ContainerConfiguration WithInstance<TExport> (string contractName, TExport instance);
public ContainerConfiguration WithInstance (Type t, object instance);
public ContainerConfiguration WithInstance (Type t, string contractName, object instance);
```

*(Thanks to @NEKIT-Boss for the examples in dotnet/runtime#18624)*

**Usage**

```
ContainerConfiguration containerConfig = new ContainerConfiguration()
    .WithAssembly(GetType().Assembly)
    .WithInstance<IExample>(new Example());

var container = containerConfig.CreateContainer();
var example = container.GetExport<IExample>();
```

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Closes #29400.

MEF looks not actively developed today, but this feature has been long required. Is it still open for change?
The implementation is directly taken and not optimal.

A usage of this is AsmDiff in arcade.

### CODE REVIEW

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs [hunk 1]
```diff
@@ -197,6 +197,54 @@ public ContainerConfiguration WithAssemblies(IEnumerable<Assembly> assemblies!!,
...
+
+        /// <summary>
+        /// Add a single instance to the container.
+        /// </summary>
+        /// <typeparam name="TExport">The type of the contract of the instance.</typeparam>
+        /// <param name="exportedInstance">The instance to add to the container.</param>
+        /// <param name="contractName">Optionally, a name that discriminates this contract from others with the same type.</param>
+        /// <param name="metadata">Optionally, a non-empty collection of named constraints that apply to the contract.</param>
+        /// <returns>A configuration object allowing configuration to continue.</returns>
+        public ContainerConfiguration WithExport<TExport>(TExport exportedInstance!!, string contractName = null, IDictionary<string, object> metadata = null)
```
Comment (tarekgh) [inline src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:219]: why using `!!` annotation here?

Comment (huoyaoyuan) [reply src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:219]: The same file is already using it. Per api review, this argument is not nullable.

Comment (tarekgh) [inline src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:219]: >string contractName = null, IDictionary<string, object> metadata = null [](http://example.com/codeflow?start=86&length=71)

The approved APIs have these parameters as nullable. any reason you didn't do so?

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs [hunk 2]
```diff
@@ -197,6 +197,54 @@ public ContainerConfiguration WithAssemblies(IEnumerable<Assembly> assemblies!!,
...
+
+        /// <summary>
+        /// Add a single instance to the container.
+        /// </summary>
+        /// <param name="contractType">The type of the contract of the instance.</param>
+        /// <param name="exportedInstance">The instance to add to the container.</param>
+        /// <param name="contractName">Optionally, a name that discriminates this contract from others with the same type.</param>
+        /// <param name="metadata">Optionally, a non-empty collection of named constraints that apply to the contract.</param>
+        /// <returns>A configuration object allowing configuration to continue.</returns>
+        public ContainerConfiguration WithExport(Type contractType!!, object exportedInstance!!, string contractName = null, IDictionary<string, object> metadata = null)
```
Comment (tarekgh) [inline src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:243]: >string contractName = null, IDictionary<string, object> metadata = null [](http://example.com/codeflow?start=97&length=71)

Same previous comment. Please make these nullable arguments as mentioned in the approval comment.

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Composition.Hosting.Core;
+
+namespace System.Composition.Hosting
+{
+    internal class InstanceExportDescriptorProvider : SinglePartExportDescriptorProvider
```
Comment (tarekgh) [inline src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs:9]: >InstanceExportDescriptorProvider [](http://example.com/codeflow?start=19&length=32)

Is it possible to get rid of SinglePartExportDescriptorProvider and have InstanceExportDescriptorProvider simply wrap a CompositionContract and _exportedInstance objects?

Would that be simpler?

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs (hunk 1 above)
Comment (huoyaoyuan) [reply src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:219]: MEF has not enabled NRT now. If `?` is used, a warning will be generated. I'm just keeping with the methods in the same file.

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs (hunk 3 above)
Comment (huoyaoyuan) [reply src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs:9]: I simply copied them from other implementations (the one used in arcade). I haven't looked around for other usages.

File: src/libraries/System.Composition.TypedParts/src/System.Composition.TypedParts.csproj [hunk 4]
```diff
@@ -18,6 +18,8 @@ System.Composition.Hosting.ContainerConfiguration</PackageDescription>
     <Compile Include="System\Composition\Debugging\ContainerConfigurationDebuggerProxy.cs" />
     <Compile Include="System\Composition\Debugging\DiscoveredPartDebuggerProxy.cs" />
     <Compile Include="System\Composition\Hosting\ContainerConfiguration.cs" />
```
Comment (tarekgh) [inline src/libraries/System.Composition.TypedParts/src/System.Composition.TypedParts.csproj:20]: @ericstj  this is a change in OOB library. The change is for .NET 7.0. Do we need to add anything to this csproj to ensure this library will release update?

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/ContainerConfiguration.cs:219]: Ok, sounds good then.

File: src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Composition.TypedParts/src/System/Composition/Hosting/InstanceExportDescriptorProvider.cs:9]: Oh, I didn't know you were copying this from another repo. I would keep this implementation then as this should be used and validated there.

File: src/libraries/System.Composition.TypedParts/src/System.Composition.TypedParts.csproj (hunk 4 above)
Comment (ericstj) [reply src/libraries/System.Composition.TypedParts/src/System.Composition.TypedParts.csproj:20]: Nothing needs to be done to ship this in `main`, we ship all libraries in the latest release.  Only during servicing do we need to enable components for shipping.

**Review by tarekgh — APPROVED** [review]

**Comment by tarekgh** [comment]
@huoyaoyuan thanks for helping to add these APIs.

**Pull request merged**